src/sources/kubernetes_logs/reconciler.rs

```diff
@@ -0,0 +1,302 @@
+
+            // Only start tailer if not already running
+            if !self.states.contains_key(&key) {
+                info!(
+                    "Starting tailer for container '{}' in pod '{}' (namespace '{}')",
+                    container_info.container_name,
+                    container_info.pod_name,
+                    container_info.namespace
+                );
+
+                self.states.insert(key, self.esb.start(container_info));
+            }
+        }
+        Ok(())
+    }
+
+    /// Clean up tailers for a deleted pod
+    async fn cleanup_pod_tailers(&mut self, pod_info: &PodInfo) {
+        for container_name in &pod_info.containers {
+            let key = ContainerKey::from((pod_info, container_name.as_str()));
+
+            if self.states.remove(&key).is_some() {
+                info!(
+                    "Cleaned up tailer for container '{}' in deleted pod '{}'",
+                    container_name, pod_info.name
+                );
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+struct EventStreamBuilder {
+    client: Client,
+    line_sender: mpsc::Sender<Vec<Line>>,
+}
+
+#[derive(Clone)]
+enum TailerState {
+    Running,
+}
+
+impl EventStreamBuilder {
+    pub fn start(&self, container_info: ContainerInfo) -> TailerState {
+        let this = self.clone();
+        tokio::spawn(async move {
+            let log_info = ContainerLogInfo::new(&container_info, Utc::now());
+            this.run_event_stream(log_info).await;
+        });
+        TailerState::Running
+    }
+
+    pub async fn run_event_stream(mut self, log_info: ContainerLogInfo<'_>) {
+        let pods: Api<Pod> =
+            Api::namespaced(self.client.clone(), &log_info.container_info.namespace);
+
+        info!(
+            "Starting streaming log tail for container '{}' in pod '{}' (namespace '{}') from timestamp {}",
+            log_info.container_info.container_name,
+            log_info.container_info.pod_name,
+            log_info.container_info.namespace,
+            log_info.log_since()
+        );
+
+        let log_params = LogParams {
+            container: Some(log_info.container_info.container_name.clone()),
+            follow: true,
+            since_time: Some(log_info.log_since()),
+            timestamps: true,
+            ..Default::default()
+        };
+
+        match pods
+            .log_stream(&log_info.container_info.pod_name, &log_params)
+            .await
+        {
+            Ok(log_stream) => {
+                info!(
+                    "Started streaming logs from container '{}' in pod '{}'",
+                    log_info.container_info.container_name, log_info.container_info.pod_name
+                );
+
+                let mut buffer = Vec::new();
+                let mut log_stream = log_stream;
+
+                // Process the stream by reading line by line
+                loop {
+                    match log_stream.read_until(b'\n', &mut buffer).await {
```

Comment:
This looks very similar to `read_until_with_max_size` from `lib/file-source-common/src/buffer.rs`. Maybe That can be used here